File: mathutil.py
"""Shared stdlib math for the NIF -> glTF pipeline. Operation order is load-bearing:
GLB output is verified byte-identical, so don't "improve" the arithmetic."""


def tf_mat(tf):
    """(translation, row-major 3x3, scale) -> 4x4 affine (row-major nested lists)."""
    t, r9, s = tf
    return [[r9[0]*s, r9[1]*s, r9[2]*s, t[0]],
            [r9[3]*s, r9[4]*s, r9[5]*s, t[1]],
            [r9[6]*s, r9[7]*s, r9[8]*s, t[2]],
            [0, 0, 0, 1]]


def mat_mul(a, b):
    return [[sum(a[i][k]*b[k][j] for k in range(4)) for j in range(4)] for i in range(4)]


def quat_from_mat3(r):
    """Row-major 3x3 -> glTF quaternion [x,y,z,w]."""
    m00, m01, m02, m10, m11, m12, m20, m21, m22 = r
    tr = m00 + m11 + m22
    if tr > 0:
        s = (tr + 1.0) ** 0.5 * 2
        w, x, y, z = 0.25 * s, (m21 - m12) / s, (m02 - m20) / s, (m10 - m01) / s
    elif m00 > m11 and m00 > m22:
        s = (1.0 + m00 - m11 - m22) ** 0.5 * 2
        w, x, y, z = (m21 - m12) / s, 0.25 * s, (m01 + m10) / s, (m02 + m20) / s
    elif m11 > m22:
        s = (1.0 + m11 - m00 - m22) ** 0.5 * 2
        w, x, y, z = (m02 - m20) / s, (m01 + m10) / s, 0.25 * s, (m12 + m21) / s
    else:
        s = (1.0 + m22 - m00 - m11) ** 0.5 * 2
        w, x, y, z = (m10 - m01) / s, (m02 + m20) / s, (m12 + m21) / s, 0.25 * s
    return [x, y, z, w]


def affine_inverse(m):
    """Inverse of a 4x4 affine matrix (row-major nested lists)."""
    a, b, c = m[0][:3], m[1][:3], m[2][:3]
    det = (a[0]*(b[1]*c[2]-b[2]*c[1]) - a[1]*(b[0]*c[2]-b[2]*c[0]) + a[2]*(b[0]*c[1]-b[1]*c[0]))
    inv3 = [
        [(b[1]*c[2]-b[2]*c[1])/det, (a[2]*c[1]-a[1]*c[2])/det, (a[1]*b[2]-a[2]*b[1])/det],
        [(b[2]*c[0]-b[0]*c[2])/det, (a[0]*c[2]-a[2]*c[0])/det, (a[2]*b[0]-a[0]*b[2])/det],
        [(b[0]*c[1]-b[1]*c[0])/det, (a[1]*c[0]-a[0]*c[1])/det, (a[0]*b[1]-a[1]*b[0])/det],
    ]
    t = [m[0][3], m[1][3], m[2][3]]
    ti = [-(inv3[i][0]*t[0] + inv3[i][1]*t[1] + inv3[i][2]*t[2]) for i in range(3)]
    return [inv3[0] + [ti[0]], inv3[1] + [ti[1]], inv3[2] + [ti[2]], [0, 0, 0, 1]]


def smooth_normals(pos, tris):
    """Area-weighted smooth normals, welding duplicate positions (UV seams)."""
    weld = {}
    canon = []
    for i, p in enumerate(pos):
        key = (round(p[0], 4), round(p[1], 4), round(p[2], 4))
        canon.append(weld.setdefault(key, i))
    acc = [[0.0, 0.0, 0.0] for _ in pos]
    for a, b, c in tris:
        ax, ay, az = pos[a]; bx, by, bz = pos[b]; cx, cy, cz = pos[c]
        ux, uy, uz = bx-ax, by-ay, bz-az
        vx, vy, vz = cx-ax, cy-ay, cz-az
        nx, ny, nz = uy*vz-uz*vy, uz*vx-ux*vz, ux*vy-uy*vx  # length = 2*area
        for i in (a, b, c):
            j = canon[i]
            acc[j][0] += nx; acc[j][1] += ny; acc[j][2] += nz
    out = []
    for i in range(len(pos)):
        x, y, z = acc[canon[i]]
        l = (x*x + y*y + z*z) ** 0.5 or 1.0
        out.append((x/l, y/l, z/l))
    return out


def bake_bind_pose(positions, normals, influences):
    """Weighted bind-pose bake: influences[vi] = [(bone_to_global 4x4, weight), ...].

    Returns (new_positions, new_normals). Accumulation follows the caller's
    influence order — keep that order stable, float sums are order-sensitive.
    """
    new_pos = list(positions)
    new_nrm = list(normals)
    for vi, infl in enumerate(influences):
        x, y, z = positions[vi]
        px = py = pz = 0.0
        nx = ny = nz = 0.0
        tw = 0.0
        for m, w in infl:
            px += w * (m[0][0]*x + m[0][1]*y + m[0][2]*z + m[0][3])
            py += w * (m[1][0]*x + m[1][1]*y + m[1][2]*z + m[1][3])
            pz += w * (m[2][0]*x + m[2][1]*y + m[2][2]*z + m[2][3])
            if normals:
                a, b, c = normals[vi]
                nx += w * (m[0][0]*a + m[0][1]*b + m[0][2]*c)
                ny += w * (m[1][0]*a + m[1][1]*b + m[1][2]*c)
                nz += w * (m[2][0]*a + m[2][1]*b + m[2][2]*c)
            tw += w
        if tw > 1e-6:
            new_pos[vi] = (px/tw, py/tw, pz/tw)
            if normals:
                l = (nx*nx + ny*ny + nz*nz) ** 0.5
                if l > 1e-6:
                    new_nrm[vi] = (nx/l, ny/l, nz/l)
    return new_pos, new_nrm


def _demo():
    ident = [[1, 0, 0, 0], [0, 1, 0, 0], [0, 0, 1, 0], [0, 0, 0, 1]]
    assert mat_mul(ident, ident) == ident
    m = tf_mat(((1, 2, 3), (0, -1, 0, 1, 0, 0, 0, 0, 1), 2.0))
    inv = affine_inverse(m)
    back = mat_mul(m, inv)
    assert all(abs(back[i][j] - ident[i][j]) < 1e-9 for i in range(4) for j in range(4)), back
    assert quat_from_mat3((1, 0, 0, 0, 1, 0, 0, 0, 1)) == [0.0, 0.0, 0.0, 1.0]
    n = smooth_normals([(0, 0, 0), (1, 0, 0), (0, 1, 0)], [(0, 1, 2)])
    assert all(abs(v[2] - 1.0) < 1e-9 for v in n), n
    # two half-weight influences at the same translation == that translation
    shift = tf_mat(((10, 0, 0), (1, 0, 0, 0, 1, 0, 0, 0, 1), 1.0))
    p, nr = bake_bind_pose([(1, 0, 0)], [(0, 0, 1)], [[(shift, 0.5), (shift, 0.5)]])
    assert p[0] == (11.0, 0.0, 0.0), p
    assert nr[0] == (0.0, 0.0, 1.0), nr
    print('mathutil ok')


if __name__ == '__main__':
    _demo()

File: internal/mathutil/mathutil.go
// Package mathutil contains the load-bearing floating-point operations used by
// the NIF to glTF pipeline. Keep evaluation and accumulation order stable:
// character GLBs are regression-checked byte for byte.
package mathutil

import (
	"math"
	"strconv"
)

type Vec3 [3]float64
type Mat3 [9]float64
type Mat4 [4][4]float64
type Transform struct {
	Translation Vec3
	Rotation    Mat3
	Scale       float64
}

// These tiny barriers prevent the compiler from fusing multiply-add sequences.
// CPython evaluates each source-level multiply and add separately; preserving
// that rounding is required for byte-identical GLB output.
//
//go:noinline
func add(a, b float64) float64 { return a + b }

//go:noinline
func sub(a, b float64) float64 { return a - b }

//go:noinline
func mul(a, b float64) float64 { return a * b }

//go:noinline
func div(a, b float64) float64 { return a / b }

func sum3(a, b, c float64) float64    { return add(add(a, b), c) }
func sum4(a, b, c, d float64) float64 { return add(add(add(a, b), c), d) }

// compensated4 mirrors CPython 3.14's CompensatedSum, which builtin sum()
// applies to the four products in mat_mul.
func compensated4(a, b, c, d float64) float64 {
	hi, lo := 0.0, 0.0
	for _, value := range [4]float64{a, b, c, d} {
		t := add(hi, value)
		if math.Abs(hi) >= math.Abs(value) {
			lo = add(lo, add(sub(hi, t), value))
		} else {
			lo = add(lo, add(sub(value, t), hi))
		}
		hi = t
	}
	if lo != 0 && !math.IsInf(lo, 0) && !math.IsNaN(lo) {
		return add(hi, lo)
	}
	return hi
}

func TransformMatrix(tf Transform) Mat4 {
	t, r, s := tf.Translation, tf.Rotation, tf.Scale
	return Mat4{
		{mul(r[0], s), mul(r[1], s), mul(r[2], s), t[0]},
		{mul(r[3], s), mul(r[4], s), mul(r[5], s), t[1]},
		{mul(r[6], s), mul(r[7], s), mul(r[8], s), t[2]},
		{0, 0, 0, 1},
	}
}

func Multiply(a, b Mat4) (out Mat4) {
	for i := range 4 {
		for j := range 4 {
			out[i][j] = compensated4(
				mul(a[i][0], b[0][j]), mul(a[i][1], b[1][j]),
				mul(a[i][2], b[2][j]), mul(a[i][3], b[3][j]),
			)
		}
	}
	return
}

func TransformPosition(m Mat4, value Vec3) Vec3 {
	x, y, z := value[0], value[1], value[2]
	return Vec3{
		sum4(mul(m[0][0], x), mul(m[0][1], y), mul(m[0][2], z), m[0][3]),
		sum4(mul(m[1][0], x), mul(m[1][1], y), mul(m[1][2], z), m[1][3]),
		sum4(mul(m[2][0], x), mul(m[2][1], y), mul(m[2][2], z), m[2][3]),
	}
}

func TransformDirection(m Mat4, value Vec3) Vec3 {
	x, y, z := value[0], value[1], value[2]
	return Vec3{
		sum3(mul(m[0][0], x), mul(m[0][1], y), mul(m[0][2], z)),
		sum3(mul(m[1][0], x), mul(m[1][1], y), mul(m[1][2], z)),
		sum3(mul(m[2][0], x), mul(m[2][1], y), mul(m[2][2], z)),
	}
}

func QuaternionFromMatrix(r Mat3) [4]float64 {
	m00, m01, m02 := r[0], r[1], r[2]
	m10, m11, m12 := r[3], r[4], r[5]
	m20, m21, m22 := r[6], r[7], r[8]
	var w, x, y, z float64
	tr := m00 + m11 + m22
	if tr > 0 {
		s := math.Sqrt(tr+1.0) * 2
		w, x, y, z = 0.25*s, (m21-m12)/s, (m02-m20)/s, (m10-m01)/s
	} else if m00 > m11 && m00 > m22 {
		s := math.Sqrt(1.0+m00-m11-m22) * 2
		w, x, y, z = (m21-m12)/s, 0.25*s, (m01+m10)/s, (m02+m20)/s
	} else if m11 > m22 {
		s := math.Sqrt(1.0+m11-m00-m22) * 2
		w, x, y, z = (m02-m20)/s, (m01+m10)/s, 0.25*s, (m12+m21)/s
	} else {
		s := math.Sqrt(1.0+m22-m00-m11) * 2
		w, x, y, z = (m10-m01)/s, (m02+m20)/s, (m12+m21)/s, 0.25*s
	}
	return [4]float64{x, y, z, w}
}

func AffineInverse(m Mat4) Mat4 {
	a, b, c := m[0], m[1], m[2]
	det := add(
		sub(
			mul(a[0], sub(mul(b[1], c[2]), mul(b[2], c[1]))),
			mul(a[1], sub(mul(b[0], c[2]), mul(b[2], c[0]))),
		),
		mul(a[2], sub(mul(b[0], c[1]), mul(b[1], c[0]))),
	)
	inv := Mat4{
		{div(sub(mul(b[1], c[2]), mul(b[2], c[1])), det), div(sub(mul(a[2], c[1]), mul(a[1], c[2])), det), div(sub(mul(a[1], b[2]), mul(a[2], b[1])), det), 0},
		{div(sub(mul(b[2], c[0]), mul(b[0], c[2])), det), div(sub(mul(a[0], c[2]), mul(a[2], c[0])), det), div(sub(mul(a[2], b[0]), mul(a[0], b[2])), det), 0},
		{div(sub(mul(b[0], c[1]), mul(b[1], c[0])), det), div(sub(mul(a[1], c[0]), mul(a[0], c[1])), det), div(sub(mul(a[0], b[1]), mul(a[1], b[0])), det), 0},
		{0, 0, 0, 1},
	}
	t := [3]float64{m[0][3], m[1][3], m[2][3]}
	for i := range 3 {
		inv[i][3] = -sum3(mul(inv[i][0], t[0]), mul(inv[i][1], t[1]), mul(inv[i][2], t[2]))
	}
	return inv
}

// Round rounds to a number of decimal places the way Python's round does:
// half-to-even against the value's exact binary expansion. Scaling by a power
// of ten and rounding instead — the obvious implementation — is off by one in
// the last digit whenever the product lands near a tie, because the scaling
// itself is inexact. The asset pipeline was ported from Python and its outputs
// are checked against committed baselines, so that last digit is load-bearing.
func Round(value float64, places int) float64 {
	rounded, err := strconv.ParseFloat(strconv.FormatFloat(value, 'f', places, 64), 64)
	if err != nil {
		return value // Only NaN and infinities, which have no decimal form.
	}
	return rounded
}

func round4(v float64) float64 { return Round(v, 4) }

func SmoothNormals(pos []Vec3, tris [][3]int) []Vec3 {
	type key [3]float64
	weld := make(map[key]int)
	canon := make([]int, len(pos))
	for i, p := range pos {
		k := key{round4(p[0]), round4(p[1]), round4(p[2])}
		j, ok := weld[k]
		if !ok {
			j = i
			weld[k] = i
		}
		canon[i] = j
	}
	acc := make([]Vec3, len(pos))
	for _, tri := range tris {
		a, b, c := pos[tri[0]], pos[tri[1]], pos[tri[2]]
		ux, uy, uz := sub(b[0], a[0]), sub(b[1], a[1]), sub(b[2], a[2])
		vx, vy, vz := sub(c[0], a[0]), sub(c[1], a[1]), sub(c[2], a[2])
		nx, ny, nz := sub(mul(uy, vz), mul(uz, vy)), sub(mul(uz, vx), mul(ux, vz)), sub(mul(ux, vy), mul(uy, vx))
		for _, vi := range tri {
			j := canon[vi]
			acc[j][0] = add(acc[j][0], nx)
			acc[j][1] = add(acc[j][1], ny)
			acc[j][2] = add(acc[j][2], nz)
		}
	}
	out := make([]Vec3, len(pos))
	for i := range pos {
		v := acc[canon[i]]
		l := math.Sqrt(sum3(mul(v[0], v[0]), mul(v[1], v[1]), mul(v[2], v[2])))
		if l == 0 {
			l = 1
		}
		out[i] = Vec3{div(v[0], l), div(v[1], l), div(v[2], l)}
	}
	return out
}

type Influence struct {
	Matrix Mat4
	Weight float64
}

func BakeBindPose(positions, normals []Vec3, influences [][]Influence) ([]Vec3, []Vec3) {
	newPos := append([]Vec3(nil), positions...)
	newNrm := append([]Vec3(nil), normals...)
	for vi, infl := range influences {
		var px, py, pz, nx, ny, nz, total float64
		for _, in := range infl {
			m, w := in.Matrix, in.Weight
			p := TransformPosition(m, positions[vi])
			px = add(px, mul(w, p[0]))
			py = add(py, mul(w, p[1]))
			pz = add(pz, mul(w, p[2]))
			if len(normals) != 0 {
				n := TransformDirection(m, normals[vi])
				nx = add(nx, mul(w, n[0]))
				ny = add(ny, mul(w, n[1]))
				nz = add(nz, mul(w, n[2]))
			}
			total = add(total, w)
		}
		if total > 1e-6 {
			newPos[vi] = Vec3{div(px, total), div(py, total), div(pz, total)}
			if len(normals) != 0 {
				l := math.Sqrt(sum3(mul(nx, nx), mul(ny, ny), mul(nz, nz)))
				if l > 1e-6 {
					newNrm[vi] = Vec3{div(nx, l), div(ny, l), div(nz, l)}
				}
			}
		}
	}
	return newPos, newNrm
}
